### PR TITLE
Allow using cabal program itself as the external setup method

### PR DESCRIPTION
**Pull request by tibbe**
This fixes issues when the version of Cabal that cabal-install was built
against differs from the one registered in the local package DB. Normally
we compile an external setup against the local Cabal library, which could
lead to failures or inconsistent results compared to using the internal
method.

### CODE REVIEW

**Comment by dcoutts** [comment]
Just a -Werror issue.

**Comment by 23Skidoo** [comment]
`getExecutablePath` is not available on GHC 7.4.2, we'll need to add it to one of the `D.C.Compat` modules.

**Comment by dcoutts** [comment]
Right, will have to do that.

**Comment by dcoutts** [comment]
BTW, @tibbe and I did this together, so the design has my +1 already. Needs some testing though.

**Comment by tibbe** [comment]
I believe I fixed the -Werror issue (and closed to old pull request due to a restriction on not using git push -f on haskell/cabal). Will look into `getExecutablePath` tonight.

**Comment by tibbe** [comment]
I've addressed the backwards compat issue with `getExecutablePath`. Turn out we already had a compat version for it in the repo.

**Comment by dcoutts** [comment]
I guess we ought to try and reproduce #2438 or #1938 and check this fixes it.

File: cabal-install/Distribution/Client/SetupWrapper.hs [hunk 1]
```diff
@@ -222,12 +223,12 @@ setupWrapper verbosity options mpkg cmd flags extraArgs = do
 --
 determineSetupMethod :: SetupScriptOptions -> BuildType -> SetupMethod
 determineSetupMethod options buildType'
-  | forceExternalSetupMethod options = externalSetupMethod
+  | buildType' == Custom             = externalSetupMethod
```
Comment (23Skidoo) [inline cabal-install/Distribution/Client/SetupWrapper.hs]: This needs a comment.

**Comment by 23Skidoo** [comment]
LGTM.

File: cabal-install/Distribution/Client/SetupWrapper.hs (hunk 1 above)
Comment (dcoutts) [inline cabal-install/Distribution/Client/SetupWrapper.hs]: Yes the logic is a bit subtle. Could put in a comment here based on my explanation here https://github.com/haskell/cabal/issues/1938#issuecomment-107456149

Comment (23Skidoo) [inline cabal-install/Distribution/Client/SetupWrapper.hs]: May be worth renaming it then.

**Comment by tibbe** [comment]
I've verified that this fixes #2438 and #1938. I've also added some comments based on @dcoutts explanation.

**Comment by 23Skidoo** [comment]
Let's merge!

**Comment by tibbe** [comment]
@dcoutts?

**Comment by dcoutts** [comment]
Yarr!

**Pull request merged**